### ISSUE TITLE
`.reset()` method not called at the end of an epoch

### ISSUE THREAD

**Issue by jul1u5** [PyTorchLightning/metrics#269]
## 🐛 Bug

In the TorchMetrics [documentation](https://torchmetrics.readthedocs.io/en/latest/pages/lightning.html):
> The .reset() method of the metric will automatically be called at the end of an epoch.

However, when I tried to use TorchMetrics, it seemed that the `.reset()` was never automatically called. 

### To Reproduce

My code looks something like this:
```python
class Model(pl.LightningModule):
    def __init__(self, ..., num_classes):
        ...
        self.valid_acc = torchmetrics.Accuracy(compute_on_step=False)

    ...

	def validation_step(self, batch, batch_idx):
		x, y = batch
        logits = self.model(x)
        loss = self.loss(logits, y)

        y_preds = logits.argmax(dim=-1)
        self.valid_acc(y_preds, y)
        self.log(f"{prefix}_loss", loss)

        return loss

    def validation_epoch_end(self, outs):
        acc = self.valid_acc.compute()
        print(f"\nValidation accuracy: {acc}")
        print(f"tp = {self.valid_acc.tp.item()}",
              f"fp = {self.valid_acc.fp.item()}",
              f"tn = {self.valid_acc.tn.item()}",
              f"fn = {self.valid_acc.fn.item()}")

        total = sum([ len(o["target"].tolist()) for o in outs ])
        print(f"total   = {total}")
        print(f"tp + fp = {self.valid_acc.tp + self.valid_acc.fp}")

        self.log("valid_acc", acc)
```

#### Output
```
Validation accuracy: 0.0
tp = 0 fp = 128 tn = 3584 fn = 128
total   = 128
tp + fp = 128

Validation accuracy: 0.3678736388683319
tp = 955 fp = 1641 tn = 85331 fn = 1641
total   = 2468
tp + fp = 2596

Validation accuracy: 0.42002370953559875
tp = 2127 fp = 2937 tn = 165823 fn = 2937
total   = 2468
tp + fp = 5064
```

### Expected behavior

`valid_acc.tp`, `valid_acc.fp` from which accuracy is calculated should be reset after each epoch, i.e., `total` should match `tp + fp`.

### Environment

I am using Kaggle Kernels.

 - PyTorch Version (e.g., 1.0): PyTorch 1.7.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: Python 3.7.9
 - CUDA/cuDNN version: 11.0
 - GPU models and configuration: Tesla P100
 - Any other relevant information:

### Additional context

I also couldn't understand what `valid_acc.tn` and `valid_acc.fn` represent. The numbers seem to be bigger than the actual number of samples.

**Comment by justusschock** [comment]
Hi @jul1u5 ,

This is intended behaviour. The [docs](https://torchmetrics.readthedocs.io/en/latest/pages/lightning.html#torchmetrics-in-pytorch-lightning) on this state, that you should log the metric object directly in order to have reset being called.

So your example would be like this:

```python
class Model(pl.LightningModule):
    def __init__(self, ..., num_classes):
        ...
        self.valid_acc = torchmetrics.Accuracy(compute_on_step=False)

    ...

    def validation_step(self, batch, batch_idx):
		x, y = batch
        logits = self.model(x)
        loss = self.loss(logits, y)

        y_preds = logits.argmax(dim=-1)
        self.valid_acc(y_preds, y)
        self.log(f"{prefix}_loss", loss)

        ### NOTE: Logging the metric object directly instead of only the computed value
        self.log('valid_acc', self.valid_acc, on_epoch=True)

        return loss

```

**Comment by tshu-w** [comment]
@justusschock So the documentation [here](https://torchmetrics.readthedocs.io/en/latest/pages/overview.html#metriccollection) is the wrong example, it took me a long time to realize the problem
